### PR TITLE
Remove pre_activation from NodeState

### PR DESCRIPTION
**Pull request by matthewbehrend**
`pre_activation` was stored in `NodeState` but never needed to persist across
inference steps. Only `LinearExplicitGrad` uses it — to evaluate
`f'(pre_activation)` for the gain-modulated error — and only within a single
forward/gradient pass. This PR removes it from the state and threads it locally
where it is still needed, cutting per-node PC inference state from 5
`(batch, width)` tensors to 4.

## Changes
- `core/types.py`: drop the `pre_activation` field and its pytree registration.
- `graph_initialization/state_initializer.py`, `nodes/base.py`: stop allocating
  / resetting `pre_activation`.
- `nodes/linear.py`: add `Linear._forward_with_preact`, returning
  `(energy, state, pre_activation)` in one pass; `forward` discards the third
  value.
- `nodes/linear_explicit_grad.py`: thread `pre_activation` from
  `_forward_with_preact` directly into
  `compute_gain_mod_error(pre_activation, error, node_info)` instead of reading
  it from state.
- `nodes/{identity,transformer,linear_residual,skip_connection,storkey_hopfield}.py`
  and example/custom nodes: drop the field from their forwards.
- `utils/dashboarding`: remove `extract_preactivation_statistics` and the
  `pre_activation` distribution/exports.
- Docs, tests, and the scaling memory analysis updated (5 → 4 tensors per node).

### CODE REVIEW

File: docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+```
+
+**File: `fabricpc/nodes/linear_explicit_grad.py`**
+
+Change the signature of `compute_gain_mod_error` so it receives `pre_activation`
+explicitly instead of reading `state.pre_activation`:
+
+```python
+@staticmethod
+def compute_gain_mod_error(
```
Comment (aprotsenko24) [inline docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md:82]: Yeah, I agree that passing it pre_activation as a local variable would be a solution for the compute_gain_mod_error. The error would have to be passed as one of the attribute from the state of the node. so state.error and it would be jnp.ndarray then, not NodeState class, as we passed the entire state object before. Cool

File: docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+`pre_activation` becomes a local variable in `forward_and_latent_grads` and
+`forward_and_weight_grads`, computed once and passed to
+`compute_gain_mod_error`. The activation API is untouched.
+
+**File: `fabricpc/nodes/linear.py`**
+
+Factor a small static helper out of `Linear.forward` so `LinearExplicitGrad` can
+reuse it:
+
+```python
```
Comment (aprotsenko24) [inline docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md:49]: Yes, having a helper method that computes pre_activation is a solution

File: fabricpc/nodes/linear.py
Comment (aprotsenko24) [inline fabricpc/nodes/linear.py:163]: The key_b is never being used later in the initialize_params() method, as the biases are set to zero -- remove splitting the key.

File: docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+batch_size = state.z_latent.shape[0]
+out_shape = node_info.shape
+flatten_input = node_info.node_config.get("flatten_input", False)
+pre_activation = Linear._compute_pre_activation(
+    params, inputs, batch_size, out_shape, flatten_input
+)
+
+# Forward pass to get state with z_mu, error, energy
+_, state = node_class.forward(params, inputs, state, node_info)
```
Comment (aprotsenko24) [inline docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md]: That is a good strategy to skip total_energy as the return value, but in several frameworks it is being assigned to total_energy and never being used, as the total_energy can be computed from the sume of the nodes' energies. Specifically, it is being done in the base.py. Harmless, but good to know.

File: docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+**File: `fabricpc/core/types.py`**
+- Lines 111-130: drop the field and its docstring entry from the `NamedTuple`.
+- Lines 190-204: drop `ns.pre_activation` from the pytree flatten lambda.
+
+### 3. Remove `pre_activation` from every constructor / `_replace` site
+
+All writes catalogued in the inventory:
+
+- `fabricpc/graph_initialization/state_initializer.py` lines ~153, ~204, ~264
```
Comment (aprotsenko24) [inline docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md:127]: Yep, removing the preactivation initialization from three initializer classes

File: scripts/storkey_hopfield_diagnostic.py [hunk 5]
```diff
@@ -576,9 +576,9 @@ def diagnostic_inference_step(params, state, clamps, structure, config):
             / (jnp.mean(jnp.linalg.norm(top_down, axis=-1)) + 1e-10)
         ),
         "tanh_saturation_frac": float(
-            jnp.mean(jnp.abs(hop_state.pre_activation) > 2.0)
+            jnp.mean(jnp.abs(hop_state.z_mu) > jnp.tanh(2.0))
```
Comment (aprotsenko24) [inline scripts/storkey_hopfield_diagnostic.py:579]: The transformation holds with tanh() activation fucntion, but if user wanted to have like a softmax function, this math would not hold:
```
|z_mu| > tanh(2.0) ⟺ |pre_activation| > 2.0 holds only because this node uses tanh (odd + strictly increasing). Don't reuse this form for non-tanh activations.
```

**Review by aprotsenko24 — APPROVED** [review]
Everything works, just put some clarifications and three possible clean-ups for the code -- replacing ```total_energy``` with ```_``` in the ```base.py```, as it is not being used. The total energy is being computed from the ```jnp.sum(<list-of-individual-node-energies>)```. Then removing redundant splitting of the key in the linear.py for the biases initialization, the biases are just being set to zero with ```jnp.zeros(...)``` and there is no randomness. So, remove ```key_b```. Lastly, I made a comment about the usgae of the activation function for the storkey_hopfield. It is tanh by default, but if user wanted to use different activation function, the transforned threshold would not work (look at comments).

File: fabricpc/nodes/linear.py
Comment (matthewbehrend) [inline fabricpc/nodes/linear.py]: Call the initializer method using the zeros init class as default

Comment (matthewbehrend) [inline fabricpc/nodes/linear.py]: bias init argument

File: scripts/storkey_hopfield_diagnostic.py (hunk 5 above)
Comment (matthewbehrend) [reply scripts/storkey_hopfield_diagnostic.py:579]: this file actually belongs better in integration tests so it can do specific work under a fixed configuration. will move it

File: fabricpc/nodes/linear.py
Comment (matthewbehrend) [reply fabricpc/nodes/linear.py:163]: Added todo for configurable bias initialization to consume the key. Out of scope for this PR.

File: scripts/storkey_hopfield_diagnostic.py (hunk 5 above)
Comment (matthewbehrend) [reply scripts/storkey_hopfield_diagnostic.py:579]: moved to scripts/

File: fabricpc/nodes/linear.py
Comment (matthewbehrend) [reply fabricpc/nodes/linear.py]: Added TODO for another PR

File: docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md (hunk 3 above)
Comment (matthewbehrend) [reply docs/dev_plans_archive/deprecate_and_remove_pre_activation_plan.md]: refactored energy summation to the base class gradient computation method where it's actually used. Nodes are no longer responsible for this duty.
